src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1944,10 +1971,173 @@ private class OffsetAndRule {
+        /// 2. Look for the data in /etc/localtime.
+        /// 3. Look for the data in c_timeZoneDirectory/localtime.
+        /// 4. Use UTC if all else fails.
+        /// </summary>
         [SecurityCritical]
-        internal static Byte[] GetLocalTzFile() {
-            // UNIXTODO
-            return null;
+        private static bool TryGetLocalTzFile(out byte[] rawData, out string id)
+        {
+            id = null;
+            string tzFilePath = null;
+            string tzVariable = GetTzEnvironmentVariable();
+
+            // GetTzEnvironmentVariable() will return non-null if the variable was set.
+            // In the case where it is set to an empty string, UTC should be used - i.e. TryGetLocalTzFile() should return false.
+
+            if (tzVariable != null)
+            {
+                if (tzVariable.Length > 0)
+                {
+                    // if tzVariable isn't an absolute path, make it relative to the system timezone directory
+                    if (tzVariable[0] != '/')
+                    {
+                        id = tzVariable;
+                        tzFilePath = Path.Combine(c_timeZoneDirectory, tzVariable);
+                    }
+                    else
+                    {
+                        tzFilePath = tzVariable;
+                    }
+
+                    if (!File.Exists(tzFilePath))
+                    {
+                        rawData = null;
+                        id = null;
+                        return false;
+                    }
+                }
+            }
+            else
+            {
+                const string localtimeFilePath = "/etc/localtime";
+                if (File.Exists(localtimeFilePath))
+                {
+                    // UNIXTODO: optimzation - see if localtimeFilePath is a symlink, and use 'readlink' to get the id
+                    tzFilePath = localtimeFilePath;
+                }
+
+                if (tzFilePath == null)
+                {
+                    string usrLocalTimeFilePath = Path.Combine(c_timeZoneDirectory, "localtime");
+                    if (File.Exists(usrLocalTimeFilePath))
+                    {
+                        tzFilePath = usrLocalTimeFilePath;
+                    }
+                }
+            }
+
+            if (string.IsNullOrEmpty(tzFilePath))
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+
+            try
+            {
+                rawData = File.ReadAllBytes(tzFilePath);
+                if (rawData != null && string.IsNullOrEmpty(id))
+                {
+                    id = FindTimeZoneId(rawData);
+                }
+
+                return rawData != null;
+            }
+            catch (IOException)
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+        }
```

Comment:
FWIW, I personally found this function's logic a bit difficult to follow, with all of the various branches and nested conditions. If it were me, I'd probably have written it something like the following (I could very well have messed this up, so if you decide to switch, please double check my logic):
```C#
private static bool TryGetLocalTzFile(out byte[] rawData, out string id)
{
    rawData = null;
    id = null;
    string tzVariable = GetTzEnvironmentVariable();

    // If the env var is null, use local time
    if (tzVariable == null)
    {
        return
            TryLoadTzFile("/etc/localtime", ref rawData, ref id) ||
            TryLoadTzFile(Path.Combine(c_timeZoneDirectory, "localtime"), ref rawData, ref id);
    }

    // If it's empty, use UTC (TryGetLocalTzFile() should return false).
    if (tzVariable.Length == 0)
    {
        return false;
    }

    // Otherwise, use the path from the env var.  If it's not absolute, make it relative 
    // to the system timezone directory
    string tzFilePath;
    if (tzVariable[0] != '/')
    {
        id = tzVariable;
        tzFilePath = Path.Combine(c_timeZoneDirectory, tzVariable);
    }
    else
    {
        tzFilePath = tzVariable;
    }
    return TryLoadTzFile(tzFilePath, ref rawData, ref id);
}

private static bool TryLoadTzFile(string tzFilePath, ref byte[] rawData, ref string id)
{
    if (File.Exists(tzFilePath))
    {
        try
        {
            rawData = File.ReadAllBytes(tzFilePath);
            if (string.IsNullOrEmpty(id))
            {
                id = FindTimeZoneId(rawData);
            }
            return true;
        }
        catch (IOException) { }
    }
    return false;
}
```